### PR TITLE
Text_similarity_reranker retriever rework to be evaluated during rewrite phase

### PR DESCRIPTION
**Pull request by pmpailis**
Now that https://github.com/elastic/elasticsearch/pull/112648 has been merged, in this PR we handle the `text_similarity_reranker` retriever next, so that it can be evaluated during the rewrite phase and be transformed to a flattened `RankDocsRetrieverBuilder`. This will enable us to support composition and many search features, including highlighting, aggregations, and collapsing.

### CODE REVIEW

**Comment by pmpailis** [comment]
@elasticmachine update branch

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankRetrieverBuilder.java [hunk 1]
```diff
@@ -32,7 +38,7 @@
 /**
  * A {@code RetrieverBuilder} for parsing and constructing a text similarity reranker retriever.
  */
-public class TextSimilarityRankRetrieverBuilder extends RetrieverBuilder {
+public class TextSimilarityRankRetrieverBuilder extends CompoundRetrieverBuilder<TextSimilarityRankRetrieverBuilder> {
 
     public static final NodeFeature TEXT_SIMILARITY_RERANKER_RETRIEVER_SUPPORTED = new NodeFeature(
         "text_similarity_reranker_retriever_supported"
```
Comment (pmpailis) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankRetrieverBuilder.java:44]: Should we introduce a new cluster feature as we did with `rrf` ?

File: x-pack/plugin/inference/src/yamlRestTest/resources/rest-api-spec/test/inference/70_text_similarity_rank_retriever.yml [hunk 2]
```diff
@@ -87,11 +87,9 @@ setup:
   - length: { hits.hits: 2 }
 
   - match: { hits.hits.0._id: "doc_2" }
-  - match: { hits.hits.0._rank: 1 }
```
Comment (pmpailis) [inline x-pack/plugin/inference/src/yamlRestTest/resources/rest-api-spec/test/inference/70_text_similarity_rank_retriever.yml:90]: `_rank` has now been removed as with `rrf`.

File: server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java [hunk 3]
```diff
@@ -464,6 +465,13 @@ private static SearchHits getHits(
                     assert shardDoc instanceof RankDoc;
                     searchHit.setRank(((RankDoc) shardDoc).rank);
                     searchHit.score(shardDoc.score);
+                    long shardAndDoc = ((long) shardDoc.shardIndex << 32) | (shardDoc.doc & 0xFFFFFFFFL);
```
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java:468]: we should have a static universal method that does this encoding.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankDoc.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+import org.apache.lucene.search.Explanation;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.search.rank.RankDoc;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Objects;
+
+public class TextSimilarityRankDoc extends RankDoc {
```
Comment (benwtrent) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankDoc.java:19]: its interesting there are no transport version changes with this.

**Review by benwtrent — COMMENTED** [review]
`TextSimilarityRankDoc.java` needs randomized serialization tests and we need to ensure whatever utilizes this named object (looks like `RankDocsQueryBuilder` and `RankDocsRetrieverBuilder`) exercise using this new named object some how. 

We don't want to miss adding this new named object to the named serialization objects and such.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankDoc.java (hunk 4 above)
Comment (pmpailis) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankDoc.java:21]: As discussed, converted this to a `VersionedNamedWriteable` and introduced a new `TransportVersion` in https://github.com/elastic/elasticsearch/pull/114085/commits/f90d278ce911119033e6957688381a8d342cbe0b

File: server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java (hunk 3 above)
Comment (pmpailis) [reply server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java]: Added in https://github.com/elastic/elasticsearch/pull/114085/commits/f90d278ce911119033e6957688381a8d342cbe0b

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankRetrieverBuilder.java (hunk 1 above)
Comment (pmpailis) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankRetrieverBuilder.java:44]: Added a new cluster feature to validate during parsing  in https://github.com/elastic/elasticsearch/pull/114085/commits/f90d278ce911119033e6957688381a8d342cbe0b

**Review by mark-vieira — APPROVED** [review]

**Comment by pmpailis** [comment]
@elasticmachine update branch

**Comment by pmpailis** [comment]
@elasticmachine update branch

**Review by benwtrent — COMMENTED** [review]
I think we are getting close, I am wondering about some of the build.gradle changes & if they are actually necessary.

File: x-pack/plugin/rank-rrf/build.gradle [hunk 5]
```diff
@@ -20,7 +20,11 @@ dependencies {
   compileOnly project(path: xpackModule('core'))
 
   testImplementation(testArtifact(project(xpackModule('core'))))
+  testImplementation(testArtifact(project(':server')))
 
   clusterModules project(xpackModule('rank-rrf'))
+  clusterModules project(xpackModule('inference'))
```
[the remark this answers is not in the document]
Comment (pmpailis) [reply x-pack/plugin/rank-rrf/build.gradle:26]: I've added this to actually exercise different semantic_reranking and rrf combinations (in `800_rrf_with_text_similarity_reranker_retriever.yml` ) . AFAICT `clusterModules` is a test-only dependency, so there shouldn't be any changes in runtime dependencies.  

However, I wasn't certain myself on what place would be the best for hosting such combined tests (needing code from both `inference` and `rank-rrf` ) so I'm really open to any alternatives :)

**Comment by pmpailis** [comment]
run elasticsearch-ci/part-1

**Review by benwtrent — APPROVED** [review]

**Pull request merged**